File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        candidates.sort()

        result, combo = list(), list()

        def dfs(index: int, combo: list[int], currSum: int) -> None:
            if currSum == target:
                result.append(combo.copy())
                return
            
            for j in range(index, len(candidates)):
                if currSum + candidates[j] > target:
                    return
                if j > index and candidates[j] == candidates[j - 1]:
                    continue
                combo.append(candidates[j])
                dfs(j + 1, combo, currSum + candidates[j])
                combo.pop()
            
        dfs(0, combo, 0)
        return result

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {

    private List<List<Integer>> result = new ArrayList<>();
    private List<Integer> combo = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] nums, int target) {

        dfs(0, nums, combo, 0, target);
        return result;
    }

    public void dfs(int index, int[] nums, List<Integer> combo, int currSum, int target) {
        if(currSum >= target || index >= nums.length) {
            if(currSum == target) {result.add(new ArrayList<>(combo));}
            return;
        }
        combo.add(nums[index]);
        dfs(index, nums, combo, currSum + nums[index], target);
        combo.removeLast();
        dfs(index + 1, nums, combo, currSum, target);
    }
}
